### DISCUSSION TITLE
Modular BBMD's communication and COV subscription reporting..

### DISCUSSION THREAD

**Discussion by eex0** [JoelBender/bacpypes#530] [Show and tell]
new modular project that utilizes bacpypes

https://github.com/eex0/BACee

I'm new, please don't be a hater ...

**Comment by bbartling** [comment]
Neat. Is that primarily for a cyber security app or more data
collections/HVAC optimization stuff?

**Reply by eex0** [reply]
if the question is: what is the code for then it depends, most any router or controllers can act as the BBMD on a network. how far down the rabbit hole you want to go?

current supported features:

Automatic Device Discovery: The code implements the Who-Is/I-Am protocol to discover BACnet devices on the network.
Supports BACnet/IP (Annex J) protocol: This is the core communication protocol used by the code.
Supports subscription to the change of value (COV) event notification: This is a primary feature of the code, and it supports both confirmed and unconfirmed COV notifications.
Supported Objects: The code can interact with the following BACnet objects:

    Analog Input
    Analog Output
    Analog Value
    Binary Input
    Binary Output
    Binary Value
    Multi-state Input
    Multi-state Output

Supports the following Objects, with most properties of the standard object types supported:
While the code provides the structure to work with these objects, it currently only reads/writes the 
'presentValue' property for COV notifications.

Supported Services: The code supports the listed BACnet services:

    Who-Is, I-Am, ReadProperty, ReadPropertyMultiple, WriteProperty (simulated), SubscribeCOV, ConfirmedCOVNotification

Supports BBMD: The client can discover and communicate with BACnet devices through a BBMD.

works in progress:
WritePropertyMultiple: The WritePropertyMultiple service is not directly implemented.
Acknowledge Event Notification in Alarm: This functionality is not present.
View monitor objects summary: to display a summary of subscribed objects and their values.

wish list:
Multi-condition triggers
Interaction among devices (Non-BACnet and BACnet): The code only interacts with BACnet devices.
A security journal.
BACnet gateway function: no gateway function to map objects between different systems.

**Comment by bbartling** [comment]
Feel free to go as far as you want. I like Alice In Wonderland : )

**Reply by eex0** [reply]
I edited my post a few times to add details, hope it helps. :)

**Comment by bbartling** [comment]
Great thanks for the update and neat project thanks for sharing. For starters if your project is new I would switch to [bacpypes3](https://github.com/JoelBender/BACpypes3). This repo is the legacy bacpypes where after python 3.12 legacy bacpypes will no longer work because of the python library `async-core` being dropped in Python 3.12 and beyond. bacpypes3 is all `asyncio` based and pretty much the same thing but it may take some refactoring.

This is neat I would totally mimic what these project goals are in BACnet network health monitoring.
- https://www.linkedin.com/posts/stevedodd_bas-smart-activity-7214998989794914304-2YDc?utm_source=share&utm_medium=member_desktop

I have a little hobby project [rest API BACnet app](https://github.com/bbartling/bacpypes3RcpServer) that is inspired off of the bacpypes3 samples for an [rpc server](https://github.com/JoelBender/BACpypes3/tree/main/samples) which has a web app front end that is sort of interesting for testing purposes on a remote network via swagger UI (comes by default with Fast API library) but it is just for hobby project in making some reads, writes, read priority arrays, etc... which can be used for troubleshooting or learning a bit how a unfamiliar BACnet network is setup.. maybe even obtaining data via rest API or whatever.

Thanks for sharing...another neat project is on legacy bacpypes is the [snap-shot](https://github.com/JoelBender/bacpypes-snapshot) project. This is a neat project that I have used few times where you can auto-scan a network completely "snap-shot" the BACnet network at whatever given moment in time that was... which could be useful if someone had to back in time and really look what going on for troubleshooting purposes beyond just typical `present-value` being logged by IoT or control system.

**Reply by eex0** [reply]
Thank You for sharing the info and links [bbartling](https://github.com/bbartling) .. I'll pop over and look at your projects tonight.

**Comment by JoelBender** [comment]
Nice project!  Here are some thoughts...
* [this line](https://github.com/eex0/BACee/blob/main/BACee.py#L19) threw me off because BACpypes3 doesn't have an IOCB module!
* [this line](https://github.com/eex0/BACee/blob/main/BACee.py#L46) you have labeled as a BBMD address when it is actually the broadcast address for the IPv4 network, given the `/24` CIDR address.  You only need address of a BBMD when you are a foreign device that is _not_ on a BACnet/IP network
* there are quite a few things in the default behavior of BACpypes3 that should be acceptable for you application (like the [DeviceInfoCache](https://github.com/JoelBender/BACpypes3/blob/main/bacpypes3/app.py#L122)) that you shouldn't need to re-implement it
* the [change of value](https://github.com/JoelBender/BACpypes3/blob/main/bacpypes3/service/cov.py#L397) API returns an async context manager that should simplify client applications like [this sample](https://github.com/JoelBender/BACpypes3/blob/main/samples/cov-client.py#L75-L88)

Thank you for your interest in using BACpypes!

**Reply by eex0** [reply]
I have been going through a bit of learning and refactoring Joel. I appreciate your looking at the project. I have taken note of your comments and will proceed to learn from and fix these issues. I decided to pull the code down today. I renamed my little project "BACeyepy". I've also decided to take my project far more serious and devote the energy into learning all I can. You Sir have been a great help!  Thank You!
